Rubberduck.Parsing/Symbols/AccessibilityCheck.cs

```diff
@@ -4,94 +4,146 @@ public static class AccessibilityCheck
-            if (IsEnclosingModule(callingModule, calleeMember))
+            if (calleeMember == null)
+            {
+                return false;
+            }    
+            else if (IsEnclosingModuleOfInstanceMember(callingModule, calleeMember))
             {
                 return true;
             }
-            var callerIsSubroutineOrProperty = callingParent.DeclarationType.HasFlag(DeclarationType.Property)
-                || callingParent.DeclarationType.HasFlag(DeclarationType.Function)
-                || callingParent.DeclarationType.HasFlag(DeclarationType.Procedure);
-            var calleeHasSameParent = callingParent.Equals(callingParent.ParentScopeDeclaration);
-            if (callerIsSubroutineOrProperty && calleeHasSameParent)
+            else if (IsLocalMemberOfTheCallingSubroutineOrProperty(callingParent, calleeMember))
             {
-                return calleeHasSameParent;
+                return true;
             }
             var memberModule = Declaration.GetModuleParent(calleeMember);
             if (IsModuleAccessible(callingProject, callingModule, memberModule))
             {
-                if (calleeMember.DeclarationType.HasFlag(DeclarationType.EnumerationMember) || calleeMember.DeclarationType.HasFlag(DeclarationType.UserDefinedTypeMember))
+                if (calleeMember.DeclarationType.HasFlag(DeclarationType.EnumerationMember) || calleeMember.DeclarationType.HasFlag(DeclarationType.UserDefinedTypeMember)) 
+                {
+                    return true;
+                }
+                else if (IsEnclosingProject(callingProject, memberModule) && IsAccessibleThroughoutTheSameProject(calleeMember))
                 {
-                    return IsValidAccessibility(calleeMember.ParentDeclaration);
+                    return true;
                 }
                 else
                 {
-                    return IsValidAccessibility(calleeMember);
+                    return HasPublicScope(calleeMember);
                 }
             }
-            return false;
+            else
+            {
+                return false;
+            }
         }
 
-        private static bool IsEnclosingModule(Declaration callingModule, Declaration calleeMember)
-        {
-            if (callingModule.Equals(calleeMember.ParentScopeDeclaration))
+            private static bool IsEnclosingModuleOfInstanceMember(Declaration callingModule, Declaration calleeMember)
             {
-                return true;
+                if (callingModule.Equals(calleeMember.ParentScopeDeclaration))
+                {
+                    return true;
+                }
+                foreach (var supertype in ClassModuleDeclaration.GetSupertypes(callingModule))
+                {
+                    if (IsEnclosingModuleOfInstanceMember(supertype, calleeMember))
+                    {
+                        return true;
+                    }
+                }
+                return false;
```

Comment:
This might be cleaner as `ClassModuleDeclaration.GetSupertypes(callingModule).Any(supertype => IsEnclosingModuleOfInstanceMember(supertype, calleeMember));`?